socketio_middleware/src/rooms.rs

```diff
@@ -33,35 +35,114 @@ impl ChannelPair {
+    }
     ROOMS.insert(room_id.to_string(), connected_sockets);
+
+    //print all sockets in the room
+    match ROOMS.get(room_id) {
+        Some(sockets) => {
+            for socket in &*sockets {
+                debug!("ROOMS: room {} containted socketid {}, sockets number = {}.", room_id, socket.sid(), sockets.len());
+            }
+        }
+        None => {
+            debug!("ROOMS: no socketid in room {}.", room_id);
+        }
+    }
 }
 
-pub fn remove_socket_from_room(room_id: &str, _sid: &str) {
+/*
+pub fn join_channel_to_room(room_id: &str, channel_pair: ChannelPair) {
+    match ROOMS.get_mut(room_id) {
+        Some(mut connected_sockets) => {
+            //check if socketid exist
+            for socket in &*connected_sockets {
+                debug!("In room {}, socket_id = {}", room_id, socket.sid());
+                if socket.sid() == channel_pair.sid() {
+                    debug!("Join channel to room {}, socket_id({}) is exist.", room_id, socket.sid());
+                    //return;
+                }
+            }
+
+            debug!("Socket_id {} insert a exist room {}", channel_pair.sid(), room_id);
+            connected_sockets.push(channel_pair);
+            ROOMS.insert(room_id.to_string(), connected_sockets.to_vec());
+        }
+        None => {
+            debug!("Socket_id {} insert a new room {}", channel_pair.sid(), room_id);
+
+            let mut connected_sockets = Vec::new();
+            connected_sockets.push(channel_pair);
+            ROOMS.insert(room_id.to_string(), connected_sockets);
+        }
+    };
+}
+*/
+
+pub fn remove_socket_from_room(room_id: &str, sid: &str) {
     let mut connected_sockets = match ROOMS.remove(room_id) {
         Some(val) => val,
         None => Vec::new(),
     };
 
-    for i in 0..connected_sockets.len() - 1 {
-        let i = connected_sockets.len() - 1 - i;
-        let socket = connected_sockets.get(i).unwrap();
-
-        if socket.sid == room_id {
+    //check if socketid exist
+    let mut i = 0;
+    for socket in &connected_sockets {
+        if socket.sid() == sid {
+            debug!("ROOMS: socketid {} leaved from room {}.", socket.sid(), room_id);
             connected_sockets.remove(i);
             break;
         }
+
+        i += 1;
     }
 
     ROOMS.insert(room_id.to_string(), connected_sockets);
+
+    //print all sockets in the room
+    match ROOMS.get(room_id) {
+        Some(sockets) => {
+            for socket in &*sockets {
+                debug!("ROOMS: room {} containted socketid {}, sockets number = {}.", room_id, socket.sid(), sockets.len());
+            }
+        }
+
+        None => {
+            debug!("ROOMS: no socketid in room {}.", room_id);
+        }
+    }
 }
 
 pub fn get_sockets_for_room(room_id: &str) -> Option<ReadGuard<String, Vec<ChannelPair>>> {
     ROOMS.get(room_id)
 }
+
+pub fn get_sockets_number_of_room(room_id: &str) -> usize {
+    match ROOMS.get(room_id) {
```

Comment:
This is a little... not rusty, can simply be

```
ROOMS.get(room_id).map(|channels| channels.len()).unwrap_or(0)
```

Reply:
Advanced rust developer ：）